Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h

```diff
@@ -0,0 +1,1886 @@
+///
+/// \details The point is on the face, meaning that all its barycentric coordinates
+///          are positive.
+///
+/// \tparam TriangleMesh A model of `FaceGraph`
+///
+/// \param fd a face of `tm`
+/// \param tm a triangulated surface mesh
+/// \param rnd optional random number generator
+///
+template <typename TriangleMesh>
+typename internal::Locate_types<TriangleMesh>::Face_location
+random_location_on_face(typename boost::graph_traits<TriangleMesh>::face_descriptor fd,
+                        const TriangleMesh& CGAL_precondition_code(tm),
+                        CGAL::Random& rnd = get_default_random())
+{
+  typedef typename internal::Locate_types<TriangleMesh>::FT               FT;
+
+  CGAL_precondition(CGAL::is_triangle_mesh(tm));
+  CGAL_precondition(fd != boost::graph_traits<TriangleMesh>::null_face());
+
+  // calling 'rnd.uniform real' with double in case FT comes from an EPECK kernel (which doesn't seem to work too well)
+  FT u(rnd.uniform_real(0., 1.));
+  FT v(rnd.uniform_real(0., double(FT(1) - u)));
+
+  return std::make_pair(fd, CGAL::make_array(u, v, FT(FT(1) - u - v)));
+}
+
+/// \brief Returns a random point over the mesh `tm`.
+///
+/// \details The returned location corresponds to a random point on a random face
+///          of the mesh. The barycentric coordinates of the point in the face
+///          are thus all positive.
+///
+/// \tparam TriangleMesh A model of `FaceGraph`
+///
+/// \param tm a triangulated surface mesh
+/// \param rnd optional random number generator
+///
+template <typename TriangleMesh>
+typename internal::Locate_types<TriangleMesh>::Face_location
+random_location_on_mesh(const TriangleMesh& tm, CGAL::Random& rnd = get_default_random())
+{
+  typedef typename boost::graph_traits<TriangleMesh>::face_descriptor     face_descriptor;
+
+  CGAL_precondition(CGAL::is_triangle_mesh(tm));
+
+  face_descriptor fd = random_face_in_mesh(tm, rnd);
+  return random_location_on_face(fd, tm, rnd);
+}
+
+/// \brief Given a `Face_location`, that is an ordered pair composed of a
+///        `boost::graph_traits<TriangleMesh>::face_descriptor` and an array
+///        of barycentric coordinates, returns a descriptor to the simplex
+///        of smallest dimension on which the point corresponding to the location lies.
+///
+/// \details In other words:
+///          - if the point lies on a vertex, this function returns a `vertex_descriptor` `v`;
+///          - if the point lies on a halfedge, this function returns a `halfedge_descriptor` `hd`
+///            (note that in that case, `loc.first == face(hd, tm)` holds).
+///          - otherwise, this function returns a `face_descriptor` `fd` (equal to `loc.first`).
+///
+/// \tparam TriangleMesh A model of `FaceGraph`
+///
+/// \pre the location corresponds to a point that is within a face of `tm`.
+/// \pre `tm` is a triangulated surface mesh.
+///
+template <typename TriangleMesh>
+typename internal::Locate_types<TriangleMesh>::descriptor_variant
```

Comment:
can't have internal in the return type of a documented function

Reply:
There are some internal in parameter types too currently... I was aiming to make a note at the top of the doc page saying "here are the types that will be used in that file", otherwise you get 5 lines-long types.

Reply:
what about a documented template alias?